sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -684,7 +684,9 @@ object LimitPushDown extends Rule[LogicalPlan] {
           left = maybePushLocalLimit(limitExpr, join.left),
           right = maybePushLocalLimit(limitExpr, join.right))
       case LeftSemi | LeftAnti if join.condition.isEmpty =>
-        join.copy(left = maybePushLocalLimit(limitExpr, join.left))
+        join.copy(
+          left = maybePushLocalLimit(limitExpr, join.left),
+          right = maybePushLocalLimit(Literal(1, IntegerType), join.right))
```

Comment:
AFAIK we have some join execution time optimization and we only fetch at most one row from the right side. @c21 is it true?

Reply:
Actually I think this PR is orthogonal to what we have did during join execution - https://github.com/apache/spark/pull/34247. During join execution, the optimization was the only keeping one row per key in hash table of SHJ. This optimizer rule here will affect join without condition, so the join would be BNLJ or `CartesianProduct`.